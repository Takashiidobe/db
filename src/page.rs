use crate::{
    row::{bytes_to_values, split_row, RowType, RowVal},
    utils::bytes_to_u32,
};
use std::{collections::BTreeMap, num::NonZeroU32};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct PageHeader {
    pub end: NonZeroU32,
    pub start: NonZeroU32,
    pub count: u32,
}

impl PageHeader {
    pub fn to_bytes(self) -> Vec<u8> {
        let mut res = self.end.get().to_le_bytes().to_vec();
        res.extend(self.start.get().to_le_bytes());
        res.extend(self.count.to_le_bytes());
        res
    }

    pub fn from_bytes(bytes: &[u8; 12]) -> Self {
        let end = NonZeroU32::new(bytes_to_u32(&bytes[0..4])).unwrap();
        let start = NonZeroU32::new(bytes_to_u32(&bytes[4..8])).unwrap();
        let count = bytes_to_u32(&bytes[8..12]);

        Self { end, start, count }
    }

    pub fn size() -> usize {
        12
    }
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Page {
    pub header: PageHeader,
    pub data: BTreeMap<NonZeroU32, Vec<RowVal>>,
    pub dirty: bool,
    pub size: usize,
    pub schema: Vec<RowType>,
}

pub const PAGE_SIZE: usize = if cfg!(feature = "small_pages") {
    56
} else {
    4096
};

impl Page {
    pub fn new(data: &[Vec<RowVal>], schema: &[RowType]) -> Self {
        let size = data
            .iter()
            .flat_map(|r| r.iter().map(|c| c.size()))
            .sum::<u16>() as usize;
        let data = BTreeMap::from_iter(data.iter().map(|row| {
            let (id, vals) = split_row(row);
            (id, vals.to_vec())
        }));

        let start = *data
            .first_key_value()
            .unwrap_or((&1.try_into().unwrap(), &vec![]))
            .0;
        let end = *data
            .last_key_value()
            .unwrap_or((&1.try_into().unwrap(), &vec![]))
            .0;

        let header = PageHeader {
            count: data.len() as u32,
            start,
            end,
        };

        Page {
            header,
            data,
            dirty: false,
            size,
            schema: schema.to_vec(),
        }
    }

    pub fn new_dirty(data: &[Vec<RowVal>], schema: &[RowType]) -> Self {
        let mut page = Page::new(data, schema);
        let page_size = data
            .iter()
            .flat_map(|r| r.iter().map(|c| c.size()))
            .sum::<u16>() as usize;
        page.dirty = true;
        page.size = page_size;
        page
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        let mut res = self.header.to_bytes();
        for (id, row) in &self.data {
            res.extend(id.get().to_le_bytes());
            for cell in row {
                res.extend(cell.clone().to_bytes());
            }
        }
        res
    }

    pub fn to_page_bytes(&self) -> Vec<u8> {
        let mut res = self.header.to_bytes();
        for (id, row) in &self.data {
            res.extend(id.get().to_le_bytes());
            for cell in row {
                res.extend(cell.clone().to_bytes());
            }
        }
        if res.len() > PAGE_SIZE {
            panic!("The page is larger than the page boundary");
        }
        let bytes_to_pad = PAGE_SIZE - res.len();
        res.extend(vec![0; bytes_to_pad]);
        res
    }

    pub fn from_bytes(bytes: &[u8], schema: &[RowType]) -> Self {
        let header_bytes: &[u8; 12] = bytes[0..12].try_into().unwrap();

        let header = PageHeader::from_bytes(header_bytes);
        let mut data = vec![];

        let mut offset = PageHeader::size();

        for _ in 0..header.count {
            let (row_val, incr) = bytes_to_values(&bytes[offset..], schema);
            data.push(row_val);
            offset += incr;
        }

        Page::new(&data, schema)
    }

    pub fn size(&self) -> usize {
        self.size
    }

    pub fn len(&self) -> usize {
        self.data.len()
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    pub fn split(&self) -> (Self, Self) {
        let len = self.len();
        let mid = len / 2;
        let vec_data: Vec<Vec<RowVal>> = self
            .data
            .clone()
            .into_iter()
            .map(|(id, row_values)| {
                let mut res = vec![RowVal::Id(id)];
                res.extend(row_values);
                res
            })
            .collect();
        let (head, tail) = vec_data.split_at(mid);

        (
            Self::new_dirty(head, &self.schema),
            Self::new_dirty(tail, &self.schema),
        )
    }

    pub fn merge(&mut self, other: Page) {
        let mut new_data = self.data.clone();
        new_data.extend(other.data);
        let vec_data: Vec<Vec<RowVal>> = new_data
            .into_iter()
            .map(|(id, row_values)| {
                let mut res = vec![RowVal::Id(id)];
                res.extend(row_values);
                res
            })
            .collect();
        *self = Self::new_dirty(&vec_data, &self.schema)
    }

    pub fn get(&self, id: NonZeroU32) -> Option<Vec<RowVal>> {
        self.data.get(&id).map(|values| values).cloned()
    }

    pub fn insert(&mut self, row: &[RowVal]) {
        let (id, values) = split_row(row);
        self.header.start = self.header.start.min(id);
        self.header.end = self.header.end.max(id);
        self.dirty = true;
        self.data.insert(id, values.to_vec());
        self.header.count = self.data.len() as u32;
    }

    pub fn remove(&mut self, id: NonZeroU32) -> Option<Vec<RowVal>> {
        match self.data.remove(&id) {
            Some(val) => {
                self.header.start = match self.data.first_key_value() {
                    Some((id, _)) => *id,
                    None => NonZeroU32::MIN,
                };
                self.header.end = match self.data.last_key_value() {
                    Some((id, _)) => *id,
                    None => NonZeroU32::MIN,
                };
                self.header.count = self.data.len() as u32;
                self.dirty = true;
                Some(val)
            }
            None => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use std::num::NonZero;

    use super::*;
    use quickcheck::{Arbitrary, Gen};
    use quickcheck_macros::quickcheck;

    impl Arbitrary for Record {
        fn arbitrary(g: &mut Gen) -> Self {
            let id = u32::arbitrary(g);
            let val = u32::arbitrary(g);
            let mut bytes: [u8; 8] = [0; 8];
            for (i, b) in id.to_le_bytes().into_iter().enumerate() {
                bytes[i] = b;
            }
            for (i, b) in val.to_le_bytes().into_iter().enumerate() {
                bytes[i + 4] = b;
            }

            Record::from_bytes(&bytes)
        }
    }

    impl Arbitrary for Page {
        fn arbitrary(g: &mut Gen) -> Self {
            let record_count = u8::arbitrary(g);
            let data: Vec<_> = (0..record_count).map(|_| Record::arbitrary(g)).collect();

            Self::new(&data)
        }
    }

    #[test]
    fn split() {
        let data = &[
            Record {
                id: NonZeroU32::new(3).unwrap(),
                val: 30,
            },
            Record {
                id: NonZeroU32::new(4).unwrap(),
                val: 40,
            },
            Record {
                id: NonZeroU32::new(1).unwrap(),
                val: 10,
            },
            Record {
                id: NonZeroU32::new(2).unwrap(),
                val: 20,
            },
        ];

        let page = Page::new(data);
        let (head, tail) = page.split();
        assert_eq!(
            head,
            Page::new_dirty(&[
                Record {
                    id: NonZero::new(1).unwrap(),
                    val: 10
                },
                Record {
                    id: NonZero::new(2).unwrap(),
                    val: 20
                },
            ])
        );
        assert_eq!(
            tail,
            Page::new_dirty(&[
                Record {
                    id: NonZero::new(3).unwrap(),
                    val: 30
                },
                Record {
                    id: NonZero::new(4).unwrap(),
                    val: 40
                }
            ])
        );
    }

    #[test]
    fn merge() {
        let data = &[
            Record {
                id: NonZeroU32::new(3).unwrap(),
                val: 30,
            },
            Record {
                id: NonZeroU32::new(4).unwrap(),
                val: 40,
            },
            Record {
                id: NonZeroU32::new(1).unwrap(),
                val: 10,
            },
            Record {
                id: NonZeroU32::new(2).unwrap(),
                val: 20,
            },
        ];
        let (head, tail) = data.split_at(data.len() / 2);

        let mut head = Page::new_dirty(head);
        head.merge(Page::new(tail));
        assert_eq!(head, Page::new_dirty(data));
    }

    #[test]
    fn get() {
        let data = vec![
            Record {
                id: NonZeroU32::new(3).unwrap(),
                val: 30,
            },
            Record {
                id: NonZeroU32::new(1).unwrap(),
                val: 10,
            },
            Record {
                id: NonZeroU32::new(2).unwrap(),
                val: 20,
            },
        ];
        let id_to_get = NonZero::new(3).unwrap();

        let page = Page::new(&data);
        let item = page.get(id_to_get);
        assert_eq!(
            item,
            Some(Record {
                id: NonZero::new(3).unwrap(),
                val: 30
            })
        );
    }

    #[test]
    fn insert() {
        let mut data = vec![
            Record {
                id: NonZeroU32::new(3).unwrap(),
                val: 30,
            },
            Record {
                id: NonZeroU32::new(1).unwrap(),
                val: 10,
            },
            Record {
                id: NonZeroU32::new(2).unwrap(),
                val: 20,
            },
        ];
        let record_to_add = Record {
            id: NonZeroU32::new(4).unwrap(),
            val: 40,
        };

        let mut head = Page::new(&data);
        head.insert(record_to_add);
        data.push(record_to_add);

        assert_eq!(head, Page::new_dirty(&data));
    }

    #[test]
    fn remove() {
        let mut data = vec![
            Record {
                id: NonZeroU32::new(3).unwrap(),
                val: 30,
            },
            Record {
                id: NonZeroU32::new(1).unwrap(),
                val: 10,
            },
            Record {
                id: NonZeroU32::new(2).unwrap(),
                val: 20,
            },
            Record {
                id: NonZeroU32::new(4).unwrap(),
                val: 40,
            },
        ];

        let id_to_remove = NonZero::new(4).unwrap();

        let mut head = Page::new(&data);
        head.remove(id_to_remove);
        data.pop();
        assert_eq!(head, Page::new_dirty(&data));
    }

    #[test]
    fn serde() {
        let data = &[
            Record {
                id: NonZeroU32::new(3).unwrap(),
                val: 30,
            },
            Record {
                id: NonZeroU32::new(4).unwrap(),
                val: 40,
            },
            Record {
                id: NonZeroU32::new(1).unwrap(),
                val: 10,
            },
            Record {
                id: NonZeroU32::new(2).unwrap(),
                val: 20,
            },
        ];

        let page = Page::new(data);

        assert_eq!(Page::from_bytes(&page.to_bytes()), page);
    }

    #[quickcheck]
    fn fuzz_page_new(records: Vec<Record>) -> bool {
        if records.len() >= u32::MAX as usize {
            return true;
        }
        let page = Page::new(&records);
        Page::from_bytes(&page.to_bytes()) == page
    }

    #[quickcheck]
    fn fuzz_page_split_merge(records: Vec<Record>) -> bool {
        if records.len() >= u32::MAX as usize {
            return true;
        }
        let page = Page::new_dirty(&records);
        let (mut head, tail) = page.split();
        head.merge(tail);
        head == page
    }
}
