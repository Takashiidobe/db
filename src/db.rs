use std::{
    collections::BTreeMap,
    fs::{File, OpenOptions},
    io::{BufWriter, Seek as _, SeekFrom, Write as _},
    num::NonZeroU32,
};

use crate::{
    row::{RowType, RowVal, Schema},
    wal::WAL,
};

use crate::page::{Page, PageHeader, PAGE_SIZE};
use indexset::{BTreeSet, Range};

#[derive(Debug)]
pub struct DB {
    pub pages: BTreeSet<(Page, Option<usize>)>,
    pub file: File,
    pub wal: WAL,
    pub epoch: u64,
    pub schema: Schema,
}

impl DB {
    pub fn new(file_name: &str, schema: &[RowType]) -> Self {
        let epoch = 1;
        let (db_file, wal_file, schema_file) = Self::setup_files(file_name, epoch);
        Self {
            file: db_file,
            pages: BTreeSet::new(),
            wal: WAL {
                file: wal_file,
                records: BTreeMap::new(),
            },
            epoch,
            schema: Schema {
                schema: schema.to_vec(),
                file: schema_file,
            },
        }
    }

    pub fn new_with_pages(
        pages: BTreeSet<(Page, Option<usize>)>,
        file_name: &str,
        schema: &[RowType],
    ) -> Self {
        let epoch = 1;
        let (db_file, wal_file, schema_file) = Self::setup_files(file_name, epoch);

        Self {
            file: db_file,
            pages,
            wal: WAL {
                file: wal_file,
                records: BTreeMap::new(),
            },
            epoch,
            schema: Schema {
                schema: schema.to_vec(),
                file: schema_file,
            },
        }
    }

    fn setup_files(file_name: &str, epoch: u64) -> (File, File, File) {
        let db_file = OpenOptions::new()
            .create(true)
            .read(true)
            .write(true)
            .open(format!("{file_name}.{epoch}.db"))
            .unwrap();
        let wal_file = OpenOptions::new()
            .create(true)
            .read(true)
            .append(true)
            .open(format!("{file_name}.{epoch}.wal"))
            .unwrap();
        let schema_file = OpenOptions::new()
            .create(true)
            .read(true)
            .write(true)
            .open(format!("{file_name}.{epoch}.schema"))
            .unwrap();
        (db_file, wal_file, schema_file)
    }

    pub fn sync(&mut self) -> bool {
        // apply all updates in wal to pages
        for (id, val) in self.wal.records.clone() {
            self.insert_to_page(id, &val);
        }

        self.serialize();
        self.wal.records.clear();
        self.wal.file.set_len(0).is_ok()
    }

    pub fn serialize(&self) {
        let mut f = BufWriter::new(&self.file);
        for (i, page) in self.pages.iter().enumerate() {
            if page.0.dirty || page.1 != Some(i) {
                let pos = SeekFrom::Start((i * PAGE_SIZE) as u64);
                let _ = f.seek(pos);
                let _ = f.write_all(&page.0.to_page_bytes());
            }
        }
        // truncation is required otherwise the page might have stale pages that have been deleted.
        let _ = self.file.set_len((self.pages.len() * PAGE_SIZE) as u64);
    }

    fn range_iter(&self, id: NonZeroU32) -> Range<(Page, Option<usize>)> {
        self.pages.range(
            (
                Page {
                    header: PageHeader {
                        end: id,
                        start: NonZeroU32::MIN,
                        count: u32::MIN,
                    },
                    dirty: false,
                    data: BTreeMap::new(),
                    size: 0,
                    schema: vec![],
                },
                None,
            )
                ..=(
                    Page {
                        header: PageHeader {
                            end: NonZeroU32::MAX,
                            start: id,
                            count: u32::MAX,
                        },
                        dirty: true,
                        data: BTreeMap::new(),
                        size: usize::MAX,
                        schema: vec![],
                    },
                    Some(usize::MAX),
                ),
        )
    }

    pub fn get(&self, id: NonZeroU32) -> Option<Vec<RowVal>> {
        // check wal first
        if let Some(val) = self.wal.get(id) {
            return Some(val);
        }

        // if not in pages, return None
        if self.pages.is_empty() {
            return None;
        }

        // otherwise, find the page where start <= id <= end
        let mut range = self.range_iter(id);

        match range.next() {
            Some(next_page) => next_page.0.get(id),
            None => None,
        }
    }

    pub fn remove(&mut self, id: NonZeroU32) -> Option<Vec<RowVal>> {
        // if in wal, remove from wal
        if let Some(val) = self.wal.remove(id) {
            return Some(val);
        }

        // if empty, return None
        if self.pages.is_empty() {
            return None;
        }

        // handle case when id is too small
        if let Some(first_page) = self.pages.first() {
            if id < first_page.0.header.start {
                return None;
            }
        }

        // handle case when id is too large
        if let Some(last_page) = self.pages.last() {
            if id > last_page.0.header.end {
                return None;
            }
        }

        // otherwise, find the page where start <= id <= end
        let mut range = self.range_iter(id);

        let next_page = range.next().unwrap();
        let mut fetched_page = next_page.clone();

        self.pages.remove(&fetched_page);
        let res = fetched_page.0.remove(id);

        // if the page still has items, readd it in
        if fetched_page.0.header.count != 0 {
            self.pages.insert(fetched_page);
        }

        res
    }

    pub fn insert(&mut self, id: NonZeroU32, val: &[RowVal]) {
        // if in wal, insert into wal
        if self.wal.insert(id, val) {
            return;
        }

        self.insert_to_page(id, val)
    }

    fn insert_to_page(&mut self, id: NonZeroU32, val: &[RowVal]) {
        let mut new_record = vec![RowVal::Id(id)];
        new_record.extend_from_slice(val);
        let row_size = val.iter().map(|x| x.size()).sum::<u16>() as usize + 4;

        // in case of an empty db
        if self.pages.is_empty() {
            let mut new_page = (Page::new_dirty(&[new_record], &self.schema.schema), None);
            new_page.0.size += row_size;
            self.pages.insert(new_page);
            return;
        }

        // handle prepend
        if let Some(first_page) = self.pages.first() {
            if id < first_page.0.header.start {
                let mut first_page = self.pages.pop_first().unwrap();
                first_page.0.size += row_size;
                first_page.0.insert(&new_record);
                self.pages.insert(first_page);

                // split page that is too big
                if let Some(first_page) = self.pages.first() {
                    if first_page.0.size() > PAGE_SIZE {
                        let (head, tail) = first_page.0.split();
                        self.pages.pop_first();
                        self.pages.insert((head, None));
                        self.pages.insert((tail, None));
                    }
                }
                return;
            }
        }

        // handle append
        if let Some(last_page) = self.pages.last() {
            if id > last_page.0.header.end {
                let mut last_page = self.pages.pop_last().unwrap();
                last_page.0.size += row_size;
                last_page.0.insert(&new_record);
                self.pages.insert(last_page);
                // split page that is too big
                if let Some(last_page) = self.pages.last() {
                    if last_page.0.size() > PAGE_SIZE {
                        let (head, tail) = last_page.0.split();
                        self.pages.pop_last();
                        self.pages.insert((head, None));
                        self.pages.insert((tail, None));
                    }
                }
                return;
            }
        }

        let mut range = self.range_iter(id);

        let next_page = range.next().unwrap();
        let mut fetched_page = next_page.clone();

        self.pages.remove(&fetched_page);
        fetched_page.0.insert(&new_record);

        if fetched_page.0.size() > PAGE_SIZE {
            let (head, tail) = fetched_page.0.split();
            self.pages.insert((head, None));
            self.pages.insert((tail, None));
        } else {
            self.pages.insert(fetched_page);
        }
    }
}

pub fn deserialize(bytes: Vec<u8>, schema: &[RowType]) -> BTreeSet<(Page, Option<usize>)> {
    assert!(bytes.len() % PAGE_SIZE == 0);

    let mut pages = vec![];

    for i in 0..(bytes.len() / PAGE_SIZE) {
        pages.push((
            Page::from_bytes(&bytes[i * PAGE_SIZE..(i + 1) * PAGE_SIZE], schema),
            Some(i),
        ));
    }

    BTreeSet::from_iter(pages)
}

impl Drop for DB {
    fn drop(&mut self) {
        self.serialize();
    }
}

#[cfg(test)]
mod tests {
    use std::{collections::HashMap, fs, num::NonZero};

    use insta::assert_yaml_snapshot as snapshot;

    use quickcheck_macros::quickcheck;

    use super::*;

    const DEFAULT_SCHEMA: &[RowType] = &[RowType::Id, RowType::U32];

    #[test]
    fn read_write() {
        let mut db = DB::new("tests/read_write", DEFAULT_SCHEMA);

        for i in 1..=5 {
            db.insert(NonZeroU32::new(i).unwrap(), &[RowVal::U32(i)]);
        }

        db.serialize();
        db.sync();

        let bytes = fs::read("tests/read_write.1.db").unwrap();

        let deserialized = deserialize(bytes, DEFAULT_SCHEMA);

        snapshot!(deserialized);
    }

    #[test]
    fn insert_loop() {
        let mut db = DB::new("tests/insert_loop", DEFAULT_SCHEMA);

        for i in 1..=510 {
            db.insert(NonZero::new(i).unwrap(), &[RowVal::U32(i)]);
        }

        db.sync();

        snapshot!(db.pages);
    }

    #[quickcheck]
    fn fuzz_db_get_insert(records: HashMap<NonZeroU32, u32>) -> bool {
        let mut db = DB::new("tests/fuzz_db_get", DEFAULT_SCHEMA);

        for (id, val) in &records {
            db.insert(*id, &[RowVal::U32(*val)]);
        }

        records
            .into_iter()
            .map(|(id, val)| db.get(id) == Some(vec![RowVal::U32(val)]))
            .all(|f| f)
    }
}
