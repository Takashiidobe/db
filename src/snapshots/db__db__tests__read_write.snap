---
source: src/db.rs
expression: deserialized
---
inner:
  - - - header:
          end: 5
          start: 1
          count: 5
        data:
          1:
            - U32: 1
          2:
            - U32: 2
          3:
            - U32: 3
          4:
            - U32: 4
          5:
            - U32: 5
        dirty: false
        size: 40
        schema:
          - Id
          - U32
      - 0
index:
  inner:
    - 1
node_capacity: 1024
len: 1
